rfd/XXXX-cloud-agent-upgrades.md

```diff
@@ -0,0 +1,318 @@
+---
+authors: Forrest Marshall (forrest@goteleport.com)
+state: draft
+---
+
+# RFD XXXX - Cloud Agent Upgrades
+
+## Required Approvers
+
+* Engineering: @klizhentas && (@russjones || @zmb3 || @rosstimothy || @espadolini)
+
+* Cloud: TBD
+
+## What
+
+A simple mechanism for automatically keeping cloud agents on a common "stable"
+install channel.
+
+## Why
+
+We would like to increase the cadence at which cloud clusters can be upgraded, and
+eliminate the significant overhead incurred by requiring users to manually upgrade agents.
+
+The existing upgrade system proposal is very complex, and many of its features
+are only relevant for niche on-prem scenarios. The majority of cloud deployments can be served
+by a far simpler system.
+
+This proposal describes a simple framework for keeping cloud agents up to date, designed with the
+intent to rely as little as possible on teleport itself for the upgrades.
+
+Because we are targeting a maximally simple system, we will assume that all cloud clusters using this
+feature track a single global "stable" channel, and that all agents are installed in one of a short
+list of supported contexts.
+
+
+## Details
+
+From the start, we will support the following scenarios:
+
+- Teleport agents deployed as `k8s` pods using a helm chart the explicitly supports this upgrade system.
+
+- Teleport agents installed via a dedicated `cloud-stable` `apt` or `yum` package.
+
+Agents deployed in one of the above manners will automically be "enrolled" in the cloud upgrade system as part of the
+installation process. Enrollment will not be coordinated at the cluster-level (i.e. agents will not check with the
+cluster to determine if upgrades should occur). It will be assumed that the latest available cloud-stable package is
+always the "correct" package to be running. This is an important divergence from how teleport versioning is typically
+handled, and has the ffect of allowing unhealthy teleport instances to attempt upgrading as a means of "fixing" themselves.
+
+The only form of cluster-level coordination around agent upgrades will be for restart timing. Since a restart is
```

Comment:
This scenario seems very compelling.

A scenario laid out during the early design meetings that we specifically want to address is one of teleport being booted from an old VM image or a machine that has been turned off for many months.  We concluded the best way to smooth things over in that scenario was to try to pull the latest version on or shortly after boot.  We also want unhealthy teleport instances to have the check run more frequently in order to let patches/rollbacks be discovered quickly.

This creates a bit of tension with the "rebooting teleport for non-teleport reasons" scenario you describe, since the best-case for the old-image/broken-teleport scenarios is to check for updates immediately, but in the scenario you lay out we really prefer to never check for updates outside of the maintenance window.

I think I've got an idea of how we can walk the line between these opposing desires:

- Change the `teleport-upgrade.service` so that it restarts teleport itself instead of having agents control their restart timing.

- By default, have the `teleport-upgrade.timer` fire at a fairly frequent rate, _but_ not so frequent that it will fire before the time we'd expect a "healthy" teleport install to be able to join the cluster (maybe every 15-30 min).

- Once healthy, have teleport export a systemd override to `teleport-upgrade.timer.d/override.conf` that blanks the monotonic timer parameters and sets an `OnCalendar` directive matching the current cloud maintenance window (e.g. `*-*-* {cloud-maintenance-window}:{jitter-minutes}:00 UTC`).

- Add an `ExecStopPost` that undoes the timer parameter override to the `teleport.service` file, so that teleport reverts to monotonic on exit... ensuring that catastrophically bugged teleport installs end up causing us to revert to using the higher-frequency monotonic timer.

- Have teleport periodically re-load and re-export the cloud maintenance window so that it can pick up on changes.

I haven't written experimental impls for this combination of options yet, but I _think_ it aught to walk the line we're trying to walk.  Lmk what you think.